### PR TITLE
[SPARK-25746][SQL][followup] do not add unnecessary If expression

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

a followup of https://github.com/apache/spark/pull/22749.

When we construct the new serializer in `ExpressionEncoder.tuple`, we don't need to add `if(isnull ...)` check for each field. They are either simple expressions that can propagate null correctly(e.g. `GetStructField(GetColumnByOrdinal(0, schema), index)`), or complex expression that already have the isnull check.

## How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @viirya

**Comment by kiszk** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 1]
```diff
@@ -124,14 +124,9 @@ object ExpressionEncoder {
...
       val input = GetStructField(GetColumnByOrdinal(0, schema), index)
-      val newDeserializer = enc.objDeserializer.transformUp {
+      enc.objDeserializer.transformUp {
         case GetColumnByOrdinal(0, _) => input
       }
-      if (schema(index).nullable) {
-        If(IsNull(input), Literal.create(null, newDeserializer.dataType), newDeserializer)
-      } else {
-        newDeserializer
-      }
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: Good catch!

**Comment by viirya** [comment]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 2]
```diff
@@ -124,14 +124,9 @@ object ExpressionEncoder {
         s"`GetColumnByOrdinal`, but there are ${getColExprs.size}")
 
       val input = GetStructField(GetColumnByOrdinal(0, schema), index)
-      val newDeserializer = enc.objDeserializer.transformUp {
+      enc.objDeserializer.transformUp {
         case GetColumnByOrdinal(0, _) => input
       }
-      if (schema(index).nullable) {
-        If(IsNull(input), Literal.create(null, newDeserializer.dataType), newDeserializer)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: oh, I see. If the child deserializer is a tuple deserializer, it is just

```scala
val deserializer =
      NewInstance(cls, childrenDeserializers, ObjectType(cls), propagateNull = false)
```

So it misses the `If(IsNull(..), null, ...)` pattern. We should wrap the `NewInstance` with `If(IsNull(..), null)` at L139.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: good catch!

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**